src/validationinterface.cpp

```diff
@@ -133,47 +138,88 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
+    auto event = [ptx, this] {
         m_internals->TransactionAddedToMempool(ptx);
-    });
+    };
+    EnqueueAndLogEvent(event, "%s: txid=%s wtxid=%s", __func__,
+                       ptx->GetHash().ToString(),
+                       ptx->GetWitnessHash().ToString());
 }
 
 void CMainSignals::TransactionRemovedFromMempool(CTransactionRef ptx, MemPoolRemovalReason reason) {
     if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
-        m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
+        auto event = [ptx, this] {
             m_internals->TransactionRemovedFromMempool(ptx);
-        });
+        };
+        EnqueueAndLogEvent(event, "%s: txid=%s wtxid=%s reason=%s", __func__,
+                           ptx->GetHash().ToString(),
+                           ptx->GetWitnessHash().ToString(),
+                           MemPoolRemovalReasonString(reason));
     }
 }
 
 void CMainSignals::BlockConnected(const std::shared_ptr<const CBlock> &pblock, const CBlockIndex *pindex, const std::shared_ptr<const std::vector<CTransactionRef>>& pvtxConflicted) {
-    m_internals->m_schedulerClient.AddToProcessQueue([pblock, pindex, pvtxConflicted, this] {
+    auto event = [pblock, pindex, pvtxConflicted, this] {
         m_internals->BlockConnected(pblock, pindex, *pvtxConflicted);
-    });
+    };
+    EnqueueAndLogEvent(event, "%s: block hash=%s block height=%d", __func__,
+                       pblock->GetHash().ToString(),
+                       pindex->nHeight);
 }
 
 void CMainSignals::BlockDisconnected(const std::shared_ptr<const CBlock> &pblock) {
-    m_internals->m_schedulerClient.AddToProcessQueue([pblock, this] {
+    auto event = [pblock, this] {
         m_internals->BlockDisconnected(pblock);
-    });
+    };
+    EnqueueAndLogEvent(event, "%s: block hash=%s", __func__,
+                       pblock->GetHash().ToString());
 }
 
 void CMainSignals::ChainStateFlushed(const CBlockLocator &locator) {
-    m_internals->m_schedulerClient.AddToProcessQueue([locator, this] {
+    auto event = [locator, this] {
         m_internals->ChainStateFlushed(locator);
-    });
+    };
+    EnqueueAndLogEvent(event, "%s: block hash=%s", __func__,
+                       locator.IsNull() ? "null" : locator.vHave.front().ToString());
 }
 
 void CMainSignals::BlockChecked(const CBlock& block, const CValidationState& state) {
+    LogEvent("%s: block hash=%s state=%s", __func__,
+             block.GetHash().ToString(), FormatStateMessage(state));
     m_internals->BlockChecked(block, state);
 }
 
 void CMainSignals::NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock> &block) {
+    LogEvent("%s: block hash=%s", __func__, block->GetHash().ToString());
     m_internals->NewPoWValidBlock(pindex, block);
 }
+
+template <typename... Args>
+void CMainSignals::EnqueueAndLogEvent(std::function<void ()> event, const char* fmt, const Args&... args) {
```

Comment:
Could `event` parameter be a const reference?

Reply:
No longer applicable as this is now a macros in 6ab424c.